hail/python/hailtop/saige/steps.py

```diff
@@ -0,0 +1,530 @@
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Tuple, Union
+
+import hail as hl
+from hailtop.aiotools.fs import AsyncFS
+import hailtop.batch as hb
+from hailtop.utils import AsyncWorkerPool
+
+from .config import CheckpointConfigMixin, JobConfigMixin
+from .constants import SaigeAnalysisType, SaigeInputDataType, SaigeTestType, saige_phenotype_to_test_type
+from .io import (
+    PlinkResourceGroup,
+    SaigeGeneGlmmResourceGroup,
+    SaigeGlmmResourceGroup,
+    SaigeSparseGRMResourceGroup,
+    SaigeGeneResultResourceGroup,
+    SaigeResultResourceGroup,
+    TextFile,
+)
+from .phenotype import Phenotype, Phenotypes
+from .variant_chunk import VariantChunk
+
+
+@dataclass
+class PrepareInputsStep(CheckpointConfigMixin):
+    async def _write_phenotype_data(
+        self,
+        fs: AsyncFS,
+        pool: AsyncWorkerPool,
+        b: hb.Batch,
+        mt: hl.MatrixTable,
+        output_file: str,
+        pheno_cols: List[str],
+    ) -> TextFile:
+        checkpointed_output = await TextFile.use_checkpoint_if_exists_and_requested(fs, pool, b, self, output_file)
+        if checkpointed_output and not self.overwrite:
+            return checkpointed_output
+        mt.cols().select(pheno_cols).export(output_file, delimiter="\t")
+        return TextFile.from_input_file(b, output_file)
+
+    async def _convert_mt_to_plink(
+        self, fs: AsyncFS, pool: AsyncWorkerPool, b: hb.Batch, mt: hl.MatrixTable, output_file: str
+    ) -> PlinkResourceGroup:
+        checkpointed_output = await PlinkResourceGroup.use_checkpoint_if_exists_and_requested(
+            fs, pool, b, self, output_file
+        )
+        if checkpointed_output and not self.overwrite:
+            return checkpointed_output
+        hl.export_plink(mt, output_file)
+        return PlinkResourceGroup.from_input_files(b, output_file)
+
+    async def call(
+        self,
+        fs: AsyncFS,
+        pool: AsyncWorkerPool,
+        b: hb.Batch,
+        mt: hl.MatrixTable,
+        phenotypes: Phenotypes,
+        working_dir: str,
+    ) -> Tuple[TextFile, PlinkResourceGroup]:
+        phenotypes_output = await self._write_phenotype_data(
+            fs, pool, b, mt, f'{working_dir}/inputs/phenotypes.tsv', phenotypes.phenotype_names
+        )
+        input_data = await self._convert_mt_to_plink(fs, pool, b, mt, f'{working_dir}/inputs/input_data')
+        return (phenotypes_output, input_data)
```

Comment:
I still feel like there are more classes and lines of code than we need to express the logic here. You're ultimately trying to get some `hailtop.batch.Resource` objects that you can use with a `job.command`, right? What do things look like if you go in this kind of direction:

```python3
async def write_plink(
    fs: AsyncFS,
    cfg: CheckpointConfigMixin,
    b: Batch,
    mt: MatrixTable,
    fname: str
) -> Resource:
    files = dict(bed=f'{fname}.bed', bim=f'{fname}.bim', fam=f'{fname}.fam')
    if cfg.overwrite or not all(await asyncio.gather(fs.exists(x) for x in files.values())):
        hl.export_plink(mt, fname)
    return b.read_input_group(**files)


async def export_tsv(
    fs: AsyncFS,
    cfg: CheckpointConfigMixin,
    b: Batch,
    ht: Table,
    fname: str
) -> Resource:
    if cfg.overwrite or not await fs.exists(fname):
        ht.export(fname, delimiter='\t')
    return b.read_input(fname)


async def prepare_inputs(
    fs: AsyncFS,
    cfg: CheckpointConfigMixin,
    b: hb.Batch,
    mt: hl.MatrixTable,
    working_dir: str,
    pheno_cols: List[str]
):
    pheno_file = f'{working_dir}/inputs/phenotypes.tsv'
    plink_file = f'{working_dir}/inputs/input_data'

    ht = mt.cols().select(pheno_cols)

    return asyncio.gather(
        export_tsv(fs, cfg, b, ht, pheno_file),
        write_plink(fs, cfg, b, mt, plink_file)
    )
```

In fact, restructuring this way reveals that we probably should have a class that contains an FS, a CheckpointConfig, and a Batch. That would save us a couple lines of code and some noise every time we call a function.

Reply:
Again, I can't quite tell what you are advocating for. Is the issue that you don't like the `PrepareInputs` step class? My intention was to make this as customizable as possible. Each "Step" in the pipeline is a class. This class must inherit from the CheckpointConfigMixin. The config for each step is present in the SaigeConfig. If a user wanted to prepare their inputs differently with custom matrix table filter functions etc., then they can inherit from the `PrepareInputs` class and override the appropriate function call (`call`). Same for all of the other steps.